src/lib/misc/roughtime/roughtime.cpp

```diff
@@ -0,0 +1,413 @@
+   return ret;
+   }
+
+std::array<uint8_t, 64> hashNode(const std::array<uint8_t, 64>& left, const std::array<uint8_t, 64>& right)
+   {
+   std::array<uint8_t, 64> ret;
+   std::unique_ptr<HashFunction> hash(HashFunction::create("SHA-512"));
+   hash->update(1);
+   hash->update(left.data(), left.size());
+   hash->update(right.data(), right.size());
+   hash->final(ret.data());
+   return ret;
+   }
+
+template<size_t N>
+std::array<uint8_t, N> vector_to_array(std::vector<uint8_t> vec)
+   {
+   if(vec.size() != N)
+      { throw Invalid_Argument("invalid vector size"); }
+   return typecast_copy<std::array<uint8_t, N>>(vec.data());
+   }
+}
+
+namespace Roughtime {
+
+Nonce::Nonce(const std::vector<uint8_t>& nonce)
+   {
+   if(nonce.size() != 64)
+      { throw Decoding_Error("invalid length for Nonce"); }
+   m_nonce = typecast_copy<std::array<uint8_t, 64>>(nonce.data());
+   }
+Nonce::Nonce(RandomNumberGenerator& rng)
+   {
+   rng.randomize(m_nonce.data(), m_nonce.size());
+   }
+
+std::array<uint8_t, request_min_size> encode_request(const Nonce& nonce)
+   {
+   std::array<uint8_t, request_min_size> buf = {2, 0, 0, 0, 64, 0, 0, 0, 'N', 'O', 'N', 'C', 'P', 'A', 'D', 0xff};
+   std::memcpy(buf.data() + 16, nonce.get_nonce().data(), nonce.get_nonce().size());
+   std::memset(buf.data() + 16 + nonce.get_nonce().size(), 0, buf.size() - 16 - nonce.get_nonce().size());
+   return buf;
+   }
+
+Response Response::from_bits(const std::vector<uint8_t>& response,
+                             const Nonce& nonce)
+   {
+   const auto response_v = split(response);
+   const auto cert = split(get_v(response_v, "CERT"));
+   const auto cert_dele = get<std::array<uint8_t, 72>>(cert, "DELE");
+   const auto cert_sig =  get<std::array<uint8_t, 64>>(cert, "SIG");
+   const auto cert_dele_v = split(cert_dele);
+   const auto srep = get_v(response_v, "SREP");
+   const auto srep_v = split(srep);
+
+   const auto cert_dele_pubk = get<std::array<uint8_t, 32>>(cert_dele_v, "PUBK");
+   const auto sig  = get<std::array<uint8_t, 64>>(response_v, "SIG");
+   if(!verify_signature(cert_dele_pubk, srep, sig))
+      { throw Invalid_Argument("response signature invalid"); }
+
+   const auto indx = get<uint32_t>(response_v, "INDX");
+   const auto path = get_v(response_v, "PATH");
+   const auto srep_root = get<std::array<uint8_t, 64>>(srep_v, "ROOT");
+   const auto size = path.size();
+   const auto levels = size/64;
+
+   if(size % 64)
+      { throw Invalid_Argument("Merkle tree path size must be multiple of 64 bytes"); }
+   if(indx >= std::exp2(levels))
+      { throw Invalid_Argument("Merkle tree path is too short"); }
+
+   auto hash = hashLeaf(nonce.get_nonce());
+   auto index = indx;
+   auto level = 0u;
+   while(level < levels)
+      {
+      hash = index&1
+             ? hashNode(typecast_copy<std::array<uint8_t, 64>>(path.data() + level*64), hash)
+             : hashNode(hash, typecast_copy<std::array<uint8_t, 64>>(path.data() + level*64));
```

Comment:
These are the only calls to hashNode. They copy the path data to another array buffer. But there is no reason to do that - just pass the raw pointers instead and have hashNode assume both inputs are 64 bytes.

(I wish we had a `span` type for things like this, but we don't get that in `std` until C++20 and I haven't gotten around to writing one for utils ...)

Reply:
Also cleaner to remove the ternary here, instead pass `index&1` to `hashNode` then it can hash left/right or right/left as needed.

Reply:
you don't want a return copy, so just hashNode(hash, path.data() + level*64, index&1), and hash is in/out parameter?

Reply:
@randombit ping

Reply:
Sorry missed this, yes in/out param is good.